pkg/controller/datadependency/handler.go

```diff
@@ -0,0 +1,900 @@
+
+	// Attempt to update the DSC status. If this fails, we must retry the whole process.
+	if err := c.updateDSCstatus(dscToUpdate); err != nil {
+		klog.Errorf("Crucial error: Failed to update status for DSC %s/%s. Retrying. Error: %v", dsc.Namespace, dsc.Name, err)
+		return err // Return error to force a requeue.
+	}
+
+	// --- 2. Best-effort: Update the status of the matched DataSource. ---
+	// Failures in this part will not fail the main binding process. They will be logged
+	// and can be reconciled later.
+
+	claimRef, err := reference.GetReference(c.Scheme, dsc)
+	if err != nil {
+		// This is a controller-side error, but not critical enough to fail the binding. Log it.
+		klog.Errorf("Could not generate ClaimRef for DSC %s/%s, DS back-references will be missing. Error: %v", dsc.Namespace, dsc.Name, err)
+		// We can proceed without the claimRef.
+		return nil
+	}
+
+	if err := c.addClaimRefToDS(matchedDS, claimRef); err != nil {
+		// Log the error but do not return it. The binding is already considered successful.
+		klog.Warningf("Best-effort failed: could not add ClaimRef to DS %s for DSC %s/%s. This may be reconciled later. Error: %v", matchedDS.Name, dsc.Namespace, dsc.Name, err)
+	}
+
+	// --- 3. Finalize ---
+	c.recorder.Eventf(dsc, corev1.EventTypeNormal, "Bound", "Successfully bound to DataSource %s", matchedDS.Name)
+	klog.V(4).Infof("Successfully completed static binding for DSC %s/%s. Main status updated, DS back-reference updated on a best-effort basis.", dsc.Namespace, dsc.Name)
+
+	return nil // The binding is successful from the scheduler's perspective.
+}
+
+func (c *DataDependencyController) dynamicBinding(dsc *v1alpha1.DataSourceClaim) error {
+	// Use plugin manager to select clusters for the data source claim
+	clusters, err := c.pluginManager.SelectClusters(context.TODO(), dsc)
+	if err != nil {
+		// Detect non-retryable "not found" scenarios from plugin (e.g., Amoro 404 ObjectNotExists)
+		// In such cases, do not requeue; just return nil to stop retrying.
+		errStr := err.Error()
+		if strings.Contains(errStr, "status 404") || strings.Contains(errStr, "ObjectNotExistsException") {
+			klog.Warningf("No data found for DSC %s/%s from plugin. Not retrying. Error: %v", dsc.Namespace, dsc.Name, err)
+			return nil
+		}
+
+		klog.Errorf("Failed to select clusters for DSC %s/%s: %v", dsc.Namespace, dsc.Name, err)
+		return err
+	}
+
+	if len(clusters) == 0 {
+		klog.V(3).Infof("No clusters selected for DSC %s/%s, keeping in Pending state", dsc.Namespace, dsc.Name)
+		return nil
+	}
+
+	// Create DataSource with selected clusters
+	// Use data source information for naming to ensure uniqueness even if DSC changes data sources
+	ds := &v1alpha1.DataSource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s-%s-%s", dsc.Spec.System, dsc.Spec.DataSourceType, dsc.Spec.DataSourceName),
+			Labels: map[string]string{
+				"datadependency.volcano.sh/system":          dsc.Spec.System,
```

Comment:
Please add these strings as constants. BTW, do we have to add these labels? What are these labels used for?